### PR TITLE
Add simple in-memory actor-realm

### PR DESCRIPTION
**Pull request by TimvdLippe**
This does not have any persistence, but just the general interface of a realm.

### CODE REVIEW

File: src/actor/ActorRealm.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { actorMixin, ValidActorMessageName } from "./Actor.js";
+
+export class ActorIDBRealm {
+  private readonly actors = new Map<ValidActorMessageName, actorMixin<any>>();
+
+  addActor<ActorName extends ValidActorMessageName>(
+    actorName: ActorName,
+    actor: actorMixin<ActorMessageType[ActorName]>
+  ) {
+    this.actors.set(actorName, actor);
```
Comment (paullewis) [inline src/actor/ActorRealm.ts]: Let's throw if we already have this actor.

File: src/actor/ActorRealm.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import { actorMixin, ValidActorMessageName } from "./Actor.js";
+
+export class ActorIDBRealm {
```
Comment (paullewis) [inline src/actor/ActorRealm.ts]: Let's also add the ability to query the names of all actors in the realm

Comment (surma) [reply src/actor/ActorRealm.ts]: This should just be called `Realm` imo

File: src/actor/ActorRealm.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  private readonly actors = new Map<ValidActorMessageName, actorMixin<any>>();
+
+  addActor<ActorName extends ValidActorMessageName>(
```
Comment (surma) [inline src/actor/ActorRealm.ts]: I’d call this `hookup`.

File: src/actor/ActorRealm.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    actorName: ActorName,
+    actor: actorMixin<ActorMessageType[ActorName]>
+  ) {
+    this.actors.set(actorName, actor);
+  }
+
+  removeActor<ActorName extends ValidActorMessageName>(actorName: ActorName) {
```
Comment (surma) [inline src/actor/ActorRealm.ts]: `hookdown`

File: src/actor/ActorRealm.ts (hunk 2 above)
Comment (surma) [reply src/actor/ActorRealm.ts]: Also +1 to Paul’s comment

File: src/actor/Actor.ts [hunk 5]
```diff
@@ -333,7 +317,7 @@ export function lookup<ActorName extends ValidActorMessageName>(
 ): ActorHandle<ActorName> {
   return {
     async send(message: ActorMessageType[ActorName]) {
-      await messageStore.pushMessage({ recipient: actorName, detail: message });
+      ACTOR_REALM.sendMessage(actorName, message);
```
Comment (surma) [inline src/actor/Actor.ts]: Which realm the actor works on should be a part of `hookup` and `lookup`, not the actor base class itself. So an actor should probably have a handle on the realm it is living in.

**Review by surma — CHANGES_REQUESTED** [review]

File: src/actor/ActorRealm.ts (hunk 3 above)
Comment (TimvdLippe) [reply src/actor/ActorRealm.ts]: I would like to keep the concepts of "adding an actor to the system" and "adding an actor to a realm" separate. E.g. a realm has actors which you can add and remove. The actor then gets hooked up on the system, without having a specific preference for the realm it is in.

File: src/actor/ActorRealm.ts (hunk 1 above)
Comment (TimvdLippe) [reply src/actor/ActorRealm.ts]: Done

File: src/actor/Actor.ts (hunk 5 above)
Comment (TimvdLippe) [reply src/actor/Actor.ts]: If you are the same process (e.g. worker, iframe, main thread), I don't think you need multiple realms. That would only complicate matters and introduce unnecessary message passing. Thus I would say that we can keep the realm local to this file and only create once for each time the file is loaded. As such, I don't think we should make realms part of `hookup` and `lookup`, but keep it in this file.

Comment (surma) [reply src/actor/Actor.ts]: Hm, we should chat about this tomorrow. To me this there is a couple of layers that all expose the same API but augment behavior. A realm is a primitive here that provides the functionality within one JavaScript realm. Then we have a layer on top by introducing bridges that makes this concept work across JavaScript realms without changing the API.

File: src/actor/Actor.ts [hunk 6]
```diff
@@ -200,7 +200,7 @@ export abstract class Actor<J> extends actorMixin(Object) {
   abstract onMessage(message: J): void;
 }
 
-const messageStore = new WatchableMessageStore("ACTOR-MESSAGES");
+const ACTOR_REALM = new ActorRealm();
```
Comment (surma) [inline src/actor/Actor.ts]: Probably named as `DEFAULT_ACTOR_REALM` or something

File: src/actor/Actor.ts [hunk 7]
```diff
@@ -12,7 +12,7 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
-import { WatchableMessageStore } from "../watchable-message-store/WatchableMessageStore.js";
+import { ActorRealm } from "./ActorRealm.js";
 
 declare global {
```
Comment (surma) [inline src/actor/Actor.ts:17]: No need to address it in this PR, but we probably need to keep in mind that with multiple (unconnected) realms it’s technically okay to have multiple actors with the same name but different messaging contracts.

File: src/actor/Actor.ts (hunk 5 above)
Comment (surma) [reply src/actor/Actor.ts]: I am trying to remember if we talked about this, but I _think_ we did. We definitely wanna support running multiple realms in the same thread, so `send()` can’t be hard-wired to the default realm.

**Review by surma — CHANGES_REQUESTED** [review]
Only one thing really, then we are good to go

**Comment by surma** [comment]
Also, wanna align with [the gist](https://gist.github.com/surma/6e1b29c1c05f46e101e9aa9f1eeabe6f) more in a separate PR or do you wanna add that to this one?

File: src/actor/Actor.ts (hunk 5 above)
Comment (surma) [reply src/actor/Actor.ts]: Nevermind, didn’t unfold the diff. All good.

**Comment by TimvdLippe** [comment]
@surma done.

**Comment by TimvdLippe** [comment]
I am no longer pursuing this PR.

**Pull request closed without merging**